### PR TITLE
Copy goal for copying files from container to host

### LINKED ISSUE #752: Need capability to copy files from running container to host
### Description

Unable to copy files in the target directory after tests have completed from container to host.

If this is already currently possible with the plugin, please let me know.

### Info

* d-m-p version : 0.20.1
* Maven version (`mvn -v`) :
```
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T08:41:47-08:00)
Maven home: /Users/axiom/tools/JavaLib/main/build/apache-maven-3.3.9
Java version: 1.8.0_92, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_92.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.10.5", arch: "x86_64", family: "mac"
```
* Docker version : 17.03.1-ce, build c6d412e

* If it's a feature request, what is your use case :

The use case I have is I'd like to copy the test results in my target directory in the post-integration-test phase from within the running container to the host.

Would be great if we could do something like this -

```xml
                 <executions>
                    <execution>
                        <id>copy</id>
                        <phase>post-integration-test</phase>
                        <configuration>
                            <fromContainerDirectory>./target/results/</fromContainerDirectory>
                            <toHostDirectory>./target/results</toHostDirectory>
                        </configuration>
                        <goals>
                            <goal>cp</goal>
                        </goals>
                    </execution>
                </executions>
```

### PR DESCRIPTION
**Pull request by mabrarov**
Copy goal for copying files & directories from container to host. Refer to issue #752.

Things which are missing due to I have few time to work on them:

- [x] Unit tests for `CopyMojo` class.
- [x] Example different than implementation of docker builder pattern and / or example copying directory (i.e. not just a single file).
- [x] Documentation for the [`copy` goal](https://github.com/mabrarov/docker-maven-plugin/blob/copy_mojo/src/main/java/io/fabric8/maven/docker/CopyMojo.java).

I appreciate any help with making this pull request complete.

Fixes #752

### CODE REVIEW

**Comment by mabrarov** [comment]
I squashed commits related to code changes for `copy` goal. Documentation and additional examples are the only things which are still pending. Though... I feel that unit tests for `CopyMojo` class could be enhanced too.

**Comment by rohanKanojia** [comment]
@mabrarov : polite ping, Could you please rebase your branch against master? Is this PR ready? Do you have some time to add examples / docs ?

**Comment by mabrarov** [comment]
@rohanKanojia,

> Could you please rebase your branch against master? 

Done.

> Is this PR ready?

Documentation is missing. It is in progress (was on hold due to the lack of time to work on it). Give me a week to complete this part of the work.

> Do you have some time to add examples

This pull request includes [it/builder test](https://github.com/mabrarov/docker-maven-plugin/blob/copy_mojo/it/builder/pom.xml) which itself is a good example. I'm looking to add an example which is different than docker builder pattern, e.g. running tests in container and copying results of tests from container to host (refer to #752). Unfortunately, I cannot promise that I'll be able to add this part of the work soon, i.e. help is appreciated.

**Comment by rhuss** [comment]
Thanks for the PR ! Unfortunately I don't have much time anymore to help except some sporadic reviews (on a month-based scale), but happy to help to integrate it as I think this covers a potential useful use case.

File: src/main/java/io/fabric8/maven/docker/config/CopyConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    @Parameter
+    private List<Entry> entries;
+
+    public List<Entry> getEntries() {
+        return entries;
+    }
+
+    public List<Properties> getEntriesAsListOfProperties() {
+        if (entries == null) {
+            return null;
```
Comment (rohanKanojia) [inline src/main/java/io/fabric8/maven/docker/config/CopyConfiguration.java:69]:
```suggestion
            return Collections.emptyList();
```

File: src/main/java/io/fabric8/maven/docker/CopyMojo.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        }
+    }
+
+    private void copyFromTemporaryContainers(final DockerAccess dockerAccess, final RunService runService,
+            final ArchiveService archiveService, final GavLabel gavLabel) throws IOException, MojoExecutionException {
+        List<ImageConfiguration> imageConfigurations = getResolvedImages();
+        for (ImageConfiguration imageConfiguration : imageConfigurations) {
+            CopyConfiguration copyConfiguration = imageConfiguration.getCopyConfiguration();
+            if (isEmpty(copyConfiguration)) {
+                continue;
```
Comment (rohanKanojia) [inline src/main/java/io/fabric8/maven/docker/CopyMojo.java:93]: I think we should log here too so that user can find out that he/she is not providing CopyConfiguration

File: src/main/java/io/fabric8/maven/docker/config/CopyConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            return this;
+        }
+
+        public CopyConfiguration build() {
+            return config;
+        }
+
+        private static List<Entry> getEntriesFromProperties(List<Properties> properties) {
+            if (properties == null) {
+                return null;
```
Comment (rohanKanojia) [inline src/main/java/io/fabric8/maven/docker/config/CopyConfiguration.java:120]:
```suggestion
                return Collections.emptyList();
```

File: it/builder/builder-image/.gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+/build/
```
Comment (rohanKanojia) [inline it/builder/builder-image/.gitignore:1]: Is this file needed?

File: it/builder/builder-image/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (rohanKanojia) [inline it/builder/builder-image/CMakeLists.txt:1]: How are we using these C related files in these integration test? Sorry, since there is no readme It's a bit difficult at a quick look.

**Comment by rohanKanojia** [comment]
Hi, I was trying to test copy goal. But somehow I'm not able to get it working. I'm trying with a configuration like this(I copied it from your integration test):
```xml
                        <plugin>
                          <groupId>io.fabric8</groupId>
                          <artifactId>docker-maven-plugin</artifactId>
                          <version>0.34-SNAPSHOT</version>
                          <executions>
                            <execution>
                              <id>copy-from-builder</id>
                              <phase>generate-resources</phase>
                              <goals>
                                <goal>start</goal>
                                <goal>copy</goal>
                                <goal>stop</goal>
                              </goals>
                              <configuration>
                                <images>
                                  <image>
                                    <name>busybox:latest</name>
                                    <copy>
                                      <entries>
                                        <entry>
                                          <containerPath>/etc/hostname</containerPath>
                                          <hostDirectory>/home/rohaan/Downloads/</hostDirectory>
                                        </entry>
                                      </entries>
                                    </copy>
                                  </image>
                                </images>
                                <createContainers>true</createContainers>
                              </configuration>
                            </execution>
                          </executions>  
                        </plugin>
```

I'm trying to copy `/etc/hostname` file from container to my Downloads folder:
```
meetup-app : $ docker run -it busybox:latest  /bin/sh
/ # cat /etc/hostname 
53369ce74e83
```

When I run `mvn clean install` I can see copy goal being executed. I can also see the file getting created, but the file is empty:
```
meetup-app : $ ls ~/Downloads/hostname 
/home/rohaan/Downloads/hostname
meetup-app : $ cat ~/Downloads/hostname 
```

Could you please review and check what I'm doing wrong?

**Comment by mabrarov** [comment]
@rohanKanojia,

Regarding [your test](https://github.com/fabric8io/docker-maven-plugin/pull/1405#issuecomment-803289305) - there is a chance that /etc/hostname file is empty at the time `copy` goal is executed, because Docker may modify /etc/hostname file after container start. Could you please provide the full maven project to reproduce the issue you faced? Is it possible to implement some pause after start of container and before invocation of `copy` goal (like [`wait` option](http://dmp.fabric8.io/#start-wait) with some timeout)?

File: it/builder/builder-image/.gitignore (hunk 4 above)
Comment (mabrarov) [reply it/builder/builder-image/.gitignore:1]: This is the directory I use for built artifacts when building related CMake project. Does it matter?

File: it/builder/builder-image/CMakeLists.txt (hunk 5 above)
Comment (mabrarov) [reply it/builder/builder-image/CMakeLists.txt:1]: Doesn't [the comment](https://github.com/fabric8io/docker-maven-plugin/blob/8e2cc922908fb7fc1cc5484a0e5f10aef3e8944f/it/builder/pom.xml#L6) (description of this integration test) in [it/builder/pom.xml](https://github.com/fabric8io/docker-maven-plugin/blob/8e2cc922908fb7fc1cc5484a0e5f10aef3e8944f/it/builder/pom.xml) answer your question? Do you feel that we need to put (duplicate?) description of the integration test in another place (I followed existing integration tests which description is placed at the top POM of the test) or / and extend it?

**Comment by rohanKanojia** [comment]
ohk, It seems to be working after removing `<createContainers>true</createContainers>` from plugin configuration. Thanks a lot

File: src/main/java/io/fabric8/maven/docker/CopyMojo.java (hunk 2 above)
Comment (mabrarov) [reply src/main/java/io/fabric8/maven/docker/CopyMojo.java:95]: Added debug log in 1b775a7. It's assumed that it's OK if copy configuration is missing, because user may have multiple images configured but wants to copy just from one of them. Please, let me know if you disagree and want to have info or warn log for this case.

File: src/main/java/io/fabric8/maven/docker/config/CopyConfiguration.java (hunk 3 above)
Comment (mabrarov) [reply src/main/java/io/fabric8/maven/docker/config/CopyConfiguration.java:120]: Refer to `io.fabric8.maven.docker.config.RunImageConfiguration#getWaitConfiguration` method. I expect that `io.fabric8.maven.docker.config.CopyConfiguration#entries` field should be `null` to follow existing design.

File: src/main/java/io/fabric8/maven/docker/config/CopyConfiguration.java (hunk 1 above)
Comment (mabrarov) [reply src/main/java/io/fabric8/maven/docker/config/CopyConfiguration.java:69]: The [same answer](https://github.com/fabric8io/docker-maven-plugin/pull/1405#discussion_r599661388) is actual here. Why I personally don;t like `null` in references representing collections, it seems to be existing design which implies `null` value if some configuration part is undefined.

File: src/main/java/io/fabric8/maven/docker/CopyMojo.java (hunk 2 above)
Comment (rohanKanojia) [reply src/main/java/io/fabric8/maven/docker/CopyMojo.java:95]: Thanks a lot. Looks good. I noticed this when I was trying out your PR and didn't configure any copyConfiguration. I was assuming that you implemented it like this:
```
mvn docker:copy -Ddocker.copy.containerPath=/src -Ddocker.copy.hostDir=/target
```

Comment (mabrarov) [reply src/main/java/io/fabric8/maven/docker/CopyMojo.java:95]: This is still possible - refer to integration test in this example:

```bash
mvn -q clean install -DskipTests \
&& mvn -q -f it/builder/builder-image/pom.xml package \
&& mvn -f it/builder/builder-image/pom.xml docker:copy \
     -Ddocker.imagePropertyConfiguration=override \
     -Ddocker.createContainers  \
     -Ddocker.copy.entries.1.containerPath=//dmp-it-builder-app.tar.gz \
     -Ddocker.copy.entries.1.hostDirectory=target \
&& tar -tvf it/builder/builder-image/target/dmp-it-builder-app.tar.gz
```

Expected output looks like:

```text
[INFO] Scanning for projects...
[INFO]
[INFO] ---------< io.fabric8.dmp.itests:dmp-it-builder-builder-image >---------
[INFO] Building dmp-it-builder-builder-image 0.34-SNAPSHOT
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- docker-maven-plugin:0.34-SNAPSHOT:copy (default-cli) @ dmp-it-builder-builder-image ---
[INFO] DOCKER> Global docker.imagePropertyConfiguration=override property activates property configuration for image
[INFO] DOCKER> Copying /dmp-it-builder-app.tar.gz from 050943faf3b8 container into D:\Users\Marat\Documents\work\java\docker-maven-plugin\it\builder\builder-image\target host directory
[INFO] Expanding: C:\Users\Marat\AppData\Local\Temp\docker-copy-7878239238909315602.tar into D:\Users\Marat\Documents\work\java\docker-maven-plugin\it\builder\builder-image\target
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.996 s
[INFO] Finished at: 2021-03-23T20:42:59+03:00
[INFO] ------------------------------------------------------------------------
-rwxr-xr-x root/root     88976 1970-01-01 03:00 hello
```

File: it/builder/builder-image/CMakeLists.txt (hunk 5 above)
Comment (rohanKanojia) [reply it/builder/builder-image/CMakeLists.txt:1]: No, I think it's okay. I just didn't notice that you had added a header there.

File: src/main/java/io/fabric8/maven/docker/CopyMojo.java (hunk 2 above)
Comment (rohanKanojia) [reply src/main/java/io/fabric8/maven/docker/CopyMojo.java:95]: Nice, Thanks a lot for your comment. I think it would be nice if we could add all these configuration modes as documentation about `docker:copy` goal in https://github.com/fabric8io/docker-maven-plugin/tree/master/src/main/asciidoc/inc

**Comment by rohanKanojia** [comment]
@mabrarov : polite ping, Do you have some time to add docs/sample? If not, I can try submitting a PR to your fork.

File: src/main/java/io/fabric8/maven/docker/CopyMojo.java (hunk 2 above)
Comment (mabrarov) [reply src/main/java/io/fabric8/maven/docker/CopyMojo.java:95]: Please note that you still need a POM with configuration of d-m-p having at least one image, because d-m-p external configuration feature can only extend / override that image configuration, but cannot create a new image configuration.

**Comment by rohanKanojia** [comment]
@mabrarov : I've created a PR for docs+sample: https://github.com/mabrarov/docker-maven-plugin/pull/1 . Would appreciate if you could review it. I think we can then proceed with merge.

**Comment by mabrarov** [comment]
Thanks to @rohanKanojia I was able to complete this PR - documentation and examples are in place now and I consider this PR ready for review and merging.

**Comment by rohanKanojia** [comment]
@mabrarov: Thanks a lot for your work! Could you please add a line to doc/changelog.md regarding this change? I think we can proceed with merge after that

I'm planning to cut 0.35.0 this week and it would be awesome if this change could make into it.

**Comment by mabrarov** [comment]
Added new `copy` goal into "doc/changelog.md". Waiting for 0.35.0 release.

**Comment by mabrarov** [comment]
@rohanKanojia,

I'm new to GitHub Actions and CircleCI, but it feels like we need to add Windows and Linux tests for (builds with) maven Wrapper, where `mvnw.cmd` (Windows) and `./mvnw` (Linux) are used instead of `mvn`.

**Comment by rohanKanojia** [comment]
@mabrarov : Hello, sorry for naive question. What's the difference in running maven goals using `mvn` as compared to these wrappers?

**Comment by rohanKanojia** [comment]
@mabrarov : I see. I don't have any problem in adding additional Github/CircleCI workflows for fixed maven wrapper. You can go ahead and create a PR for this if you have time. If not, you can create an issue and I'll try to pick it up in the coming holidays.

**Comment by mabrarov** [comment]
@rohanKanojia,

Here is the new feature request about Maven Wrapper: #1450

**Comment by rohanKanojia** [comment]
@mabrarov : I've created https://github.com/fabric8io/docker-maven-plugin/pull/1451 for adding github action workflows for maven wrappers. Would appreciate if you could review it..

**Review by rohanKanojia — APPROVED** [review]

**Comment by rohanKanojia** [comment]
@mabrarov : Thanks a lot for your PR! Could you please resolve this minor conflict? I think we can proceed with merge then
